lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,270 @@
+                  " argument(s) not allowed"_err_en_US});
+        }
+      }
+      // C1139: call to impure procedure
+      if (name->symbol && !isPure(name->symbol->attrs())) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"call to impure subroutine in DO"
+                                         " CONCURRENT not allowed"_err_en_US});
+      }
+      if (name->symbol && fromScope(*name->symbol, "IEEE_EXCEPTIONS"s)) {
+        auto upperName{parser::ToUpperCaseLetters(name->ToString())};
+        if (upperName == "IEEE_GET_FLAG"s) {
+          messages_.Say(charBlock_,
+              parser::MessageFormattedText{
+                  "IEEE_GET_FLAG not allowed in DO CONCURRENT"_err_en_US});
+        } else if (upperName == "IEEE_SET_HALTING_MODE"s) {
+          messages_.Say(charBlock_,
+              parser::MessageFormattedText{"IEEE_SET_HALTING_MODE not allowed"
+                                           " in DO CONCURRENT"_err_en_US});
+        } else if (upperName == "IEEE_GET_HALTING_MODE"s) {
+          messages_.Say(charBlock_,
+              parser::MessageFormattedText{"IEEE_GET_HALTING_MODE not allowed"
+                                           " in DO CONCURRENT"_err_en_US});
+        }
+      }
+    } else {
+      // C1139: this a procedure component
+      auto &component{std::get<parser::ProcComponentRef>(procedureDesignator.u)
+                          .v.thing.component};
+      if (component.symbol && !isPure(component.symbol->attrs())) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"call to impure subroutine in DO"
+                                         " CONCURRENT not allowed"_err_en_US});
+      }
+    }
+  }
+
+  // 11.1.7.5
+  void Post(const parser::IoControlSpec &ioControlSpec) {
+    if (auto *charExpr{
+            std::get_if<parser::IoControlSpec::CharExpr>(&ioControlSpec.u)}) {
+      if (std::get<parser::IoControlSpec::CharExpr::Kind>(charExpr->t) ==
+          parser::IoControlSpec::CharExpr::Kind::Advance) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "ADVANCE specifier not allowed in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+private:
+  bool anyObjectIsCoarray() { return false; }  // placeholder
+  bool anyObjectIsPolymorphic() { return false; }  // placeholder
+  bool EndTDeallocatesCoarray() { return false; }  // placeholder
+  bool isPure(Attrs &attrs) {
+    return attrs.HasAny({Attr::PURE}) ||
+        (attrs.HasAny({Attr::ELEMENTAL}) && !attrs.HasAny({Attr::IMPURE}));
+  }
+  bool fromScope(const Symbol &symbol, const std::string &moduleName) {
+    if (symbol.scope() && symbol.scope()->IsModule()) {
+      if (parser::ToUpperCaseLetters(
+              symbol.scope()->symbol()->name().ToString()) == moduleName) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  std::set<parser::Label> labels_;
+  parser::CharBlock charBlock_;
```

Comment:
I think it would be better for `charBlocks_` to have a name that indicates its purpose, e.g. `currStmt_`.

Reply:
ok